pallet/account-migration/src/lib.rs

```diff
@@ -16,174 +16,167 @@
-		pub migrated_accounts: Vec<(AccountId32, Balance)>,
-	}
+	/// [`frame_system::Account`] data.
+	#[pallet::storage]
+	#[pallet::getter(fn account_of)]
+	pub type Accounts<T: Config> =
+		StorageMap<_, Identity, AccountId32, AccountInfo<Index, AccountData<Balance>>>;
 
-	#[pallet::genesis_build]
-	impl<T: Config> GenesisBuild<T> for GenesisConfig {
-		fn build(&self) {
-			self.migrated_accounts.iter().for_each(|(account, amount)| {
-				Balances::<T>::insert(account, amount);
-			});
-		}
-	}
+	// TODO: identity storages
+	// TODO: proxy storages
+	// TODO: staking storages
+	// TODO: vesting storages
 
 	#[pallet::call]
-	impl<T: Config> Pallet<T>
-	where
-		AccountIdOf<T>: From<H160>,
-		BalanceOf<T>: From<Balance>,
-	{
-		// since signature and chain_id verification is done in `validate_unsigned`
-		// we can skip doing it here again.
-		// TODO: update weight
+	impl<T: Config> Pallet<T> {
+		/// Migrate all the account data under the `from` to `to`.
 		#[pallet::weight(0)]
-		pub fn claim_to(
+		pub fn migrate(
 			origin: OriginFor<T>,
-			_chain_id: u64,
-			old_pub_key: AccountId32,
-			new_pub_key: H160,
-			_sig: Signature,
+			from: AccountId32,
+			to: AccountId20,
+			_signature: Signature,
 		) -> DispatchResult {
 			ensure_none(origin)?;
 
-			let Some(amount) = Balances::<T>::take(&old_pub_key) else  {
-				return Err(Error::<T>::AccountNotExist.into());
-			};
+			let account = <Accounts<T>>::take(from)
+				.ok_or("[pallet::account-migration] already checked in `pre_dispatch`; qed")?;
+
+			<frame_system::Account<T>>::insert(to, account);
 
-			<T as pallet::Config>::Currency::deposit_creating(&new_pub_key.into(), amount.into());
-			Self::deposit_event(Event::Claim { old_pub_key, new_pub_key, amount });
+			Self::deposit_event(Event::Migrated { from, to });
 
 			Ok(())
 		}
 	}
 	#[pallet::validate_unsigned]
-	impl<T: Config> ValidateUnsigned for Pallet<T>
-	where
-		AccountIdOf<T>: From<H160>,
-		BalanceOf<T>: From<Balance>,
-	{
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
 		type Call = Call<T>;
 
 		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
-			let Call::claim_to { chain_id, old_pub_key, new_pub_key, sig } = call else {
+			// The migration destination was already taken by someone.
+			const E_ACCOUNT_ALREADY_EXISTED: u8 = 0;
+			// The migration source was not exist.
+			const E_ACCOUNT_NOT_FOUND: u8 = 1;
+			// Invalid signature.
+			const E_INVALID_SIGNATURE: u8 = 1;
+
+			let Call::migrate { from, to, signature } = call else {
 				return InvalidTransaction::Call.into();
 			};
 
-			if *chain_id != <T as pallet_evm::Config>::ChainId::get() {
-				return InvalidTransaction::BadProof.into();
+			if !<Accounts<T>>::contains_key(from) {
+				return InvalidTransaction::Custom(E_ACCOUNT_NOT_FOUND).into();
 			}
-			if !Balances::<T>::contains_key(old_pub_key) {
-				return InvalidTransaction::BadSigner.into();
+			if <frame_system::Account<T>>::contains_key(to) {
```

Comment:
This check enforces the `to` must be a fresh account(never existed in this chain). Is this expected?

Seems weird, the users are not able to migrate the old account to an already-used account in the new chain.

Reply:
Otherwise, it will override the existing data. Then we could make the logic much simpler.

Reply:
What do you think about just changing the `AccountData` to `Balance` or something does not contain nonce, providers, like what I did before? In that way,  without such a `to` restriction. But I am not sure if we need the `reserved` and `misc_frozen` data here.

Reply:
@AsceticBear I think it make sense for me to make the logic simpler by adding the requirement the migration target account should not exists.



This requirement is reasonable and can keep it simple, well worth it